src/transformStateWithClones.js

```diff
@@ -8,6 +8,33 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
```

Comment:
Improvement:

**1. Avoid Mutating currentState in Loops:**
While currentState is reassigned in each iteration, mutating its properties (like in the removeProperties case) can lead to unexpected behavior. Instead, create a new object during every update.

**2. Avoid Direct delete in Loops:**
Directly deleting properties can sometimes lead to performance issues. Use destructuring or Object.fromEntries for better clarity and immutability.

```
      case 'removeProperties':
        // Remove specified properties immutably
        const keysToRemove = new Set(action.keysToRemove);
        currentState = Object.fromEntries(
          Object.entries(currentState).filter(([key]) => !keysToRemove.has(key))
        );
        break;
```